src/bootstrap/setup.rs

```diff
@@ -107,6 +108,60 @@ pub fn setup(src_path: &Path, profile: Profile) {
     let include_path = profile.include_path(src_path);
     println!("`x.py` will now use the configuration at {}", include_path.display());
 
+    let build = TargetSelection::from_user(&env!("BUILD_TRIPLE"));
+    let stage_path = ["build", build.rustc_target_arg(), "stage1"].join("/");
+
+    let rustup_installed = Command::new("rustup")
+        .arg("--version")
+        .stdout(std::process::Stdio::null())
+        .output()
+        .map_or(false, |output| output.status.success());
+
+    let stage_dir_exists = match fs::create_dir(&stage_path[..]) {
+        Ok(_) => true,
+        Err(_) => Path::new(&stage_path[..]).exists(),
+    };
+
+    if rustup_installed && stage_dir_exists {
+        // Let's see if we have already linked the stage1 toolchain.
+        let toolchain_list = Command::new("rustup")
+            .args(&["toolchain", "list"])
+            .stdout(std::process::Stdio::piped())
+            .output();
+
+        if let Ok(toolchain_list) = toolchain_list {
+            let links = String::from_utf8_lossy(&toolchain_list.stdout);
+            if !links.contains("stage1") {
+                let toolchain_link = Command::new("rustup")
+                    .stdout(std::process::Stdio::null())
+                    .args(&["toolchain", "link", "stage1", &stage_path[..]])
+                    .output();
+                if toolchain_link.is_err() || !toolchain_link.unwrap().status.success() {
+                    println!("`rustup` failed to link stage 1 build to `stage1` toolchain");
+                    println!(
+                        "To manually link stage 1 build to `stage1` toolchain, run:\n
+                        `rustup toolchain link stage1 {}`",
+                        &stage_path[..]
+                    );
+                } else {
+                    println!(
+                        "Added `stage1` rustup toolchain; try `cargo +stage1 build` on a separate rust project to run a newly-built toolchain"
+                    );
+                }
+            } else {
+                println!(
+                    "`stage1` toolchain already linked; not attempting to link `stage1` toolchain",
+                );
+            }
+        } else {
+            println!(
+                "`rustup` failed to list current toolchains; not attempting to link `stage1` toolchain"
+            );
+        }
+    } else if profile != Profile::User {
+        println!("`rustup` is not installed; cannot link `stage1` toolchain");
+    }
```

Comment:
This long list of `else` blocks makes it hard to tell which error corresponds with which check IMO - could you instead use a separate functions and early returns so it's more clear? Something like this:
```rust
    maybe_create_rustup_toolchain();
// ...
}

fn maybe_create_rustup_toolchain() {
     if !rustup_installed || !stage_dir_exists {
        // ...
        return
     }
     // ...
}
```